.github/workflows/ci-cd.yml

```diff
@@ -0,0 +1,754 @@
+    - name: Log stdlib OpenSSL version
+      run: >-
+        python -c
+        "import ssl; print('\nOPENSSL_VERSION: '
+        + ssl.OPENSSL_VERSION + '\nOPENSSL_VERSION_INFO: '
+        + repr(ssl.OPENSSL_VERSION_INFO)
+        + '\nOPENSSL_VERSION_NUMBER: '
+        + repr(ssl.OPENSSL_VERSION_NUMBER))"
+
+    - name: Check rst
+      run: |
+        python setup.py check --restructuredtext
+
+    # this ensures our database is ready. typically by the time the preparations have completed its first start logic.
+    # unfortunately we need this hacky workaround as GitHub Actions service containers can't reference data from our repo.
+    - name: Prepare mysql
+      run: |
+        # ensure server is started up
+        while :
+        do
+            sleep 1
+            mysql -h127.0.0.1 -uroot "-p$MYSQL_ROOT_PASSWORD" -e 'select version()' && break
+        done
+
+        # inject tls configuration
+        docker container stop mysqld
+        docker container cp "${{ github.workspace }}/tests/ssl_resources/ssl" mysqld:/etc/mysql/ssl
+        docker container cp "${{ github.workspace }}/tests/ssl_resources/tls.cnf" mysqld:/etc/mysql/conf.d/aiomysql-tls.cnf
+
+        # use custom socket path
+        # we need to ensure that the socket path is writable for the user running the DB process in the container
+        sudo chmod 0777 /tmp/run-${{ join(matrix.db, '-') }}
+
+        # mysql 5.7 container overrides the socket path in /etc/mysql/mysql.conf.d/mysqld.cnf
+        if [ "${{ join(matrix.db, '-') }}" = "mysql-5.7" ]
+        then
+          docker container cp "${{ github.workspace }}/tests/ssl_resources/socket.cnf" mysqld:/etc/mysql/mysql.conf.d/zz-aiomysql-socket.cnf
+        else
+          docker container cp "${{ github.workspace }}/tests/ssl_resources/socket.cnf" mysqld:/etc/mysql/conf.d/aiomysql-socket.cnf
+        fi
+
+        docker container start mysqld
+
+        # ensure server is started up
+        while :
+        do
+            sleep 1
+            mysql -h127.0.0.1 -uroot "-p$MYSQL_ROOT_PASSWORD" -e 'select version()' && break
+        done
+
+        mysql -h127.0.0.1 -uroot "-p$MYSQL_ROOT_PASSWORD" -e "SET GLOBAL local_infile=on"
+
+    - name: Run tests
+      run: |
+        # timeout ensures a more or less clean stop by sending a KeyboardInterrupt which will still provide useful logs
+        timeout --preserve-status --signal=INT --verbose 5m \
+          pytest --capture=no --verbosity 2 --cov-report term --cov-report xml --cov aiomysql --cov tests ./tests --mysql-unix-socket "unix-${{ join(matrix.db, '') }}=/tmp/run-${{ join(matrix.db, '-') }}/mysql.sock" --mysql-address "tcp-${{ join(matrix.db, '') }}=127.0.0.1:3306"
+      env:
+        PYTHONUNBUFFERED: 1
+      timeout-minutes: 6
+
+    - name: Upload coverage
+      uses: codecov/codecov-action@v2.1.0
+      with:
+        file: ./coverage.xml
+        flags: "${{ matrix.os }}_${{ matrix.py }}_${{ join(matrix.db, '-') }}"
```

Comment:
It's best to use comma-separated flags. Codecov allows ticking each one separately and combining to choose different aspects of coverage to show.

Reply:
I'm not sure if I'm missing something here or it's just not implemented correctly in the `cheroot` repo?

https://github.com/cherrypy/cheroot/blob/81f9c687c264d0ec2c1b8bfa3b59caffe739db72/.github/workflows/ci-cd.yml#L689-L693

https://github.com/cherrypy/cheroot/runs/5147091997?check_suite_focus=true#step:19:4
logs
```yaml
with:
  flags: GHA, Linux, 3.4, cheroot-8.6.1.dev388+ge1fb8e54-py2.py3-none-any.whl
```
the upload URL then only contains `&flags=GHA`:
https://github.com/cherrypy/cheroot/runs/5147091997?check_suite_focus=true#step:19:38
and on codecov this only shows a single flag: https://app.codecov.io/gh/cherrypy/cheroot/compare/491/overview

while on the current configuration I can see a lot of flags on codecov:
https://app.codecov.io/gh/aio-libs/aiomysql/compare/706

Reply:
according to [codecov-action's arg docs](https://github.com/marketplace/actions/codecov#arguments)

> Flag the upload to group coverage metrics (unittests, uitests, etc.). Multiple flags are separated by a comma (ui,chrome)

considering https://docs.codecov.com/docs/flags#one-to-one-relationship-of-flags-to-uploads documents that flags should be 1:1, my understanding would be that the same flag should not be used by multiple uploads, unless it replaces the previous upload's data.

the example of `(ui,chrome)` does however make it sound reasonable to also use e.g. `(ui,firefox)`.

Reply:
The action definitely does pass the separated flags to the codecov executable though: https://github.com/cherrypy/cheroot/runs/5147091997?check_suite_focus=true#step:19:21

Reply:
_TL;DR this would work today._

That judgement wasn't right. The better workaround would've been removing spaces between flags, and having commas like `flag1,flag2,flag3`.
The case with `flag1, flag2, flag3` wasn't being processed correctly back when you made this PR. It was a legit Codecov bug. Recently, I've found more, like stripping trailing zeroes when flags could be interpreted as floats.
The said bugs have been fixed in the action and the uploader projects. Some by me (not my first time fixing their bugs!)
Here are the changes, if you're curious: https://github.com/codecov/uploader/issues/876 / https://github.com/codecov/uploader/pull/878 / https://github.com/codecov/codecov-action/pull/791.

This now works: https://github.com/cherrypy/cheroot/blob/70a5980/.github/workflows/ci-cd.yml#L713-L724. I've migrated to having prefixed flags after discovering that bug with float-ish inputs, before the fix got released. Though I think it's even better this way so I kept the approach.
Another nice bit is extracting the exact Python version number that `actions/setup-python` installed ­— it's better then relying on the matrix factors (which may even contain chars that aren't acceptable for flags). See https://github.com/cherrypy/cheroot/blob/70a5980/.github/workflows/ci-cd.yml#L555.